src/main/java/com/iota/iri/service/TipsManager.java

```diff
@@ -272,70 +302,161 @@ static long capSum(long a, long b, long max) {
         return a+b;
     }
 
-    void serialUpdateRatings(final Set<Hash> visitedHashes, final Hash txHash, final Map<Hash, Long> ratings, final Set<Hash> analyzedTips, final Hash extraTip) throws Exception {
-        Stack<Hash> hashesToRate = new Stack<>();
-        hashesToRate.push(txHash);
-        Hash currentHash;
-        boolean addedBack;
-        while(!hashesToRate.empty()) {
-            currentHash = hashesToRate.pop();
-            TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, currentHash);
-            addedBack = false;
-            Set<Hash> approvers = transactionViewModel.getApprovers(tangle).getHashes();
-            for(Hash approver : approvers) {
-                if(ratings.get(approver) == null && !approver.equals(currentHash)) {
-                    if(!addedBack) {
-                        addedBack = true;
-                        hashesToRate.push(currentHash);
+    /**
+     * Updates the cumulative weight of txs.
+     * A cumulative weight of each tx is 1 + the number of ancestors it has.
+     *
+     * See https://github.com/alongalky/iota-docs/blob/master/cumulative.md
+     *
+     *
+     * @param myApprovedHashes the current hashes of the snapshot at the time of calculation
+     * @param currentTxHash the transaction from where the analysis starts
+     * @param confirmLeftBehind if true attempt to give more weight to previously
+     *                          unconfirmed txs
+     * @throws Exception if there is a problem accessing the db
+     */
+    Map<Buffer, Integer> calculateCumulativeWeight(Set<Hash> myApprovedHashes, Hash currentTxHash, boolean confirmLeftBehind,
+            Set<Hash> analyzedTips) throws Exception {
+        log.info("Start calculating cw starting with tx hash {}", currentTxHash);
+        log.debug("Start topological sort");
+        long start = System.currentTimeMillis();
+        LinkedHashSet<Hash> txHashesToRate = sortTransactionsInTopologicalOrder(currentTxHash);
+        log.debug("Subtangle size: {}", txHashesToRate.size());
+        log.debug("Topological sort done. Start traversing on txs in order and calculate weight");
+        Map<Buffer, Integer> cumulativeWeights = calculateCwInOrder(txHashesToRate, myApprovedHashes, confirmLeftBehind,
+                analyzedTips);
+        log.debug("Cumulative weights calculation done in {} ms", System.currentTimeMillis() - start);
+        return cumulativeWeights;
+    }
+
+    private LinkedHashSet<Hash>  sortTransactionsInTopologicalOrder(Hash startTx) throws Exception {
+        LinkedHashSet<Hash> sortedTxs = new LinkedHashSet<>();
+        Set<Hash> temporary = new HashSet<>();
+        Deque<Hash> stack = new ArrayDeque<>();
+        Map<Hash, Collection<Hash>> txToDirectApprovers = new HashMap<>();
+
+        stack.push(startTx);
+        while (CollectionUtils.isNotEmpty(stack)) {
```

Comment:
I think `!stack.isEmpty()` is cleaner, and more consistent with the rest of the calls

Reply:
The idea be`CollectionUtils` is to be defensive and fall for `NullPointerException`.

You can say that in this specific code segment `stack` will never be `null`, and even if it can the above line (`stack.push()`) will throw an exception.

However, the reason I use it is purely out of habit, which I believe to be a good one.
If wherever it is fine to treat a `null` collection like an empty colllection one will use a null safe method then we will get less pesky null pointer exceptions.